### PR TITLE
feat: add desktopentry class

### PR DESCRIPTION
**Pull request by ComixHe**
Log:

### CODE REVIEW

File: src/desktopentry.cpp [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+            }
+        }else {
+            unescapedStr.append(c);
+        }
+    }
+
+    return unescapedStr;
+}
+
+QString DesktopEntry::Value::toString(const QString& key) const noexcept
```
Comment (black-desk) [inline src/desktopentry.cpp:166]: should this be an optional<QString>?

File: src/desktopentry.cpp [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+{
```
Comment (black-desk) [inline src/desktopentry.cpp:167]: I think the `key` is not that useful in this method.

When user try to parse a localestring as string, I think this toString method should just return the one without any \[].

And a user should not get the localed value of this key.

File: src/desktopentry.cpp [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+}
+
+QString DesktopEntry::Value::toLocaleString(const QLocale &locale) const noexcept
+{
+    for (auto it = this->constKeyValueBegin(); it != this->constKeyValueEnd(); ++it) {
+        auto [a, b] = *it;
+        if (QLocale{a}.name() == locale.name())
+            return unescape(b);
+    }
+    return (*this)["default"];
```
Comment (black-desk) [inline src/desktopentry.cpp:185]: use defaultKeyStr

File: src/include/desktopentry.h [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+// SPDX-License-Identifier: LGPL-3.0-or-later
+
+#include <QString>
+#include <QMap>
+#include <QVariant>
+#include <QMetaType>
+#include <QDebug>
+#include <QLocale>
+
+constexpr static auto defaultKeyStr = "default";
```
Comment (black-desk) [inline src/include/desktopentry.h:12]: should this be just a empty string?

File: src/desktopentry.cpp [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+        currentGroup->insert(keyStr, {{valueKey, valueStr}});
+    } else {
+        if (auto v = cur->find(valueKey); v == cur->end()) {
+            cur->insert(valueKey, valueStr);
+        }
+    }
+    return ParseError::NoError;
+}
+
+DesktopEntry::ParseError DesktopEntry::parse(const QString &desktopFileName) noexcept
```
Comment (black-desk) [inline src/desktopentry.cpp:54]: > Desktop entry files should have the .desktop extension, except for files of Type Directory which should have the .directory extension. Determining file type on basis of extension makes determining the file type very easy and quick. When no file extension is present, the desktop system should fall back to recognition via "magic detection".

I think we should refer to the spec, or just copy the spec into comments, when we try to decide something is invalid.

File: src/desktopentry.cpp [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+        qWarning() << "file isn't a desktop file";
+        return ParseError::MismatchedFile;
+    }
+    auto XDGDataDirs = qgetenv("XDG_DATA_DIRS").split(':');
+
+    qDebug() << "Current Application Dirs:" << XDGDataDirs;
+    for (const auto &d : XDGDataDirs) {
+        auto dirPath = QDir::cleanPath(d);
+        QDirIterator it{dirPath,{desktopFileName},QDir::AllEntries | QDir::NoSymLinks | QDir::NoDotAndDotDot, QDirIterator::Subdirectories};
+        QStringList fileList;
```
Comment (black-desk) [inline src/desktopentry.cpp:66]: unused

File: src/desktopentry.cpp [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+}
+
+QMap<QString, DesktopEntry::Value> DesktopEntry::group(const QString &key) const noexcept
+{
+    if (auto group = entryMap.find(key); group != entryMap.cend())
+        return *group;
+    return {};
+}
+
+QString DesktopEntry::Value::unescape(const QString& str) const noexcept
```
Comment (black-desk) [inline src/desktopentry.cpp:125]:
```cpp
QString DesktopEntry::Value::unescape(const QString &str) const noexcept
{
    QString unescapedStr;
    for (qsizetype i = 0; i < str.size(); ++i) {
        auto c = str.at(i);
        if (c != '\\') {
            unescapedStr.append(c);
            continue;
        }

        switch (str.at(i + 1).toLatin1()) {
            default:
                unescapedStr.append(c);
                break;
            case 'n':
                unescapedStr.append('\n');
                ++i;
                break;
            case 't':
                unescapedStr.append('\t');
                ++i;
                break;
            case 'r':
                unescapedStr.append('\r');
                ++i;
                break;
            case '\\':
                unescapedStr.append('\\');
                ++i;
                break;
            case ';':
                unescapedStr.append(';');
                ++i;
                break;
            case 's':
                unescapedStr.append(' ');
                ++i;
                break;
        }
    }

    return unescapedStr;
}
```

File: src/desktopentry.cpp [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+    return unescapedStr;
+}
+
+QString DesktopEntry::Value::toString(const QString& key) const noexcept
+{
+    auto str = this->find(key);
+    if (str == this->end())
+        return {};
+    auto unescapedStr = unescape(*str);
+    if (unescapedStr.contains(QRegularExpression{"\\p{Cc}"}) and
```
Comment (black-desk) [inline src/desktopentry.cpp:172]: give these constant strings a name.

File: src/desktopentry.cpp (hunk 3 above)
Comment (ComixHe) [reply src/desktopentry.cpp:185]: done

File: src/desktopentry.cpp (hunk 6 above)
Comment (ComixHe) [reply src/desktopentry.cpp:66]: done

File: src/desktopentry.cpp (hunk 7 above)
Comment (ComixHe) [reply src/desktopentry.cpp:125]: done

File: src/desktopentry.cpp (hunk 8 above)
Comment (ComixHe) [reply src/desktopentry.cpp:172]: done

File: src/include/desktopentry.h (hunk 4 above)
Comment (ComixHe) [reply src/include/desktopentry.h:12]: I think the empty string may confuse the user

File: src/desktopentry.cpp (hunk 2 above)
Comment (ComixHe) [reply src/desktopentry.cpp:167]: this key has a default value

File: src/include/desktopentry.h [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+    auto parserGroupHeader(const QString &str) noexcept;
+    ParseError parseEntry(const QString &str, decltype(m_entryMap)::iterator &currentGroup) noexcept;
+};
+
+struct DesktopFile
+{
+    const QString filePath;
+    const QString desktopId;
+
+    static std::unique_ptr<DesktopFile> searchDesktopFile(const QString &desktopFilePath, ParseError& err) noexcept;
```
Comment (black-desk) [inline src/include/desktopentry.h:61]: maybe use optional?

File: src/desktopentry.cpp [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+ParseError DesktopEntry::parseEntry(const QString &str, decltype(m_entryMap)::iterator& currentGroup) noexcept
+{
+    if (str.startsWith("#"))
+        return ParseError::NoError;
+    auto splitCharIndex = str.indexOf("=");
+    auto keyStr = str.first(splitCharIndex).trimmed();
+    auto valueStr = str.sliced(splitCharIndex + 1).trimmed();
+    QString key, valueKey{defaultKeyStr};
+
+    constexpr auto validKeyChars = "[^0-9a-zA-Z-]";
```
Comment (black-desk) [inline src/desktopentry.cpp:34]: invalidKeyChars
```suggestion
    constexpr auto invalidKeyChars = "[^0-9a-zA-Z-]";
```

**Pull request merged**